x/fibre/blob.go

```diff
@@ -0,0 +1,307 @@
+
+	extendedData *rsema1d.ExtendedData
+	commitment   Commitment
+	rlcOrig      []field.GF128
+
+	// holds meta fields about the blob
+	header blobHeaderV0
+	// data holds the decoded original data (without header).
+	data []byte
+}
+
+// NewBlob creates a new [Blob] instance by encoding the data.
+// It takes the data and a [BlobConfig].
+// The data is prefixed with a header containing the blob version and data size.
+func NewBlob(data []byte, cfg BlobConfig) (d *Blob, err error) {
+	if len(data) == 0 {
+		return nil, fmt.Errorf("data cannot be empty")
+	}
+	if len(data) > cfg.MaxBlobSize {
+		return nil, fmt.Errorf("%w: data size %d exceeds maximum %d", ErrBlobTooLarge, len(data), cfg.MaxBlobSize)
+	}
+
+	d = &Blob{
+		cfg:    cfg,
+		header: newBlobHeaderV0(len(data)),
+		data:   data,
+	}
+
+	rows := d.header.encodeToRows(data, cfg)
+	d.extendedData, d.commitment, d.rlcOrig, err = rsema1d.Encode(rows, &rsema1d.Config{
+		K:           cfg.OriginalRows,
+		N:           cfg.ParityRows,
+		RowSize:     len(rows[0]),
+		WorkerCount: cfg.CodingWorkers,
+	})
+	if err != nil {
+		return nil, fmt.Errorf("encoding data: %w", err)
+	}
+
+	return d, nil
+}
+
+// Commitment returns the commitment to the blob.
+func (d *Blob) Commitment() Commitment {
+	return d.commitment
+}
+
+// RLCOrig returns the original RLC coefficients.
+func (d *Blob) RLCOrig() []field.GF128 {
+	return d.rlcOrig
+}
+
+// RowSize returns the size of each row in bytes.
+// Returns 0 if no original data available to determine row size.
+func (d *Blob) RowSize() int {
+	if len(d.data) == 0 {
+		return 0
+	}
+
+	return d.header.calculateRowSize(len(d.data), d.cfg)
+}
+
+// DataSize returns the size of the original data (without header) by reading from the blob header.
+// Returns 0 if no original data available to determine its size.
+func (d *Blob) DataSize() int {
+	return len(d.data)
+}
+
+// Size returns the total size of the blob including the header overhead.
+// Returns 0 if no original data available to determine blob size.
+func (d *Blob) Size() int {
+	dataSize := d.DataSize()
+	if dataSize == 0 {
+		return 0
+	}
+	return blobHeaderSize + dataSize
+}
+
+// Data returns the cached original data (without header).
+// Returns nil if the data has not been decoded yet (call Reconstruct first for received blobs).
+func (d *Blob) Data() []byte {
+	return d.data
+}
+
+// Row returns the [rsema1d.RowProof] for the given index from the extended data.
+func (d *Blob) Row(index int) (*rsema1d.RowProof, error) {
+	if d.extendedData == nil {
+		return nil, fmt.Errorf("no extended data available")
+	}
+
+	return d.extendedData.GenerateRowProof(index)
+}
+
+const (
+	// uint32SizeBytes is the size of a uint32 in bytes.
+	uint32SizeBytes = 4
+	// uint8SizeBytes is the size of a uint8 in bytes.
+	uint8SizeBytes = 1
+	// blobHeaderSize is the size of the blob header in bytes.
+	// Format: 1 byte version (uint8) + 4 bytes blob size (uint32)
+	blobHeaderSize = uint8SizeBytes + uint32SizeBytes
```

Comment:
Was this meant to be encoding constants? Those should probably reflect it by being named not for uint length, but for encoding meaning:
```suggestion
	// blobLengthSizeBytes is the size of a blob length encoding in bytes.
	blobLengthSizeBytes = 4
	// blobVersionSizeBytes is the size of a blob version encoding in bytes.
	blobVersionSizeBytes = 1
	// blobHeaderSize is the size of the blob header in bytes.
	// Format: 1 byte version (uint8) + 4 bytes blob size (uint32)
	blobHeaderSize = blobVersionSizeBytes + dataLengthSizeBytes
```

Reply:
Agree, this can be better. 

I am going for 

```
const (
	// blobVersionLen is the length of the version field in bytes.
	blobVersionLen = 1
	// blobDataSizeLen is the length of the data size field in bytes.
	blobDataSizeLen = 4
	// blobHeaderLen is the total length of the blob header in bytes.
	// Format: 1 byte version + 4 bytes data size
	blobHeaderLen = blobVersionLen + blobDataSizeLen
)
```

Using Len is more idiomatic